Loop/View Controllers/CarbEntryViewController.swift

```diff
@@ -216,23 +216,25 @@ final class CarbEntryViewController: LoopChartsTableViewController, Identifiable
     private var foodKeyboard: EmojiInputController!
     
     private func updateDisplayAccurateCarbEntryWarning() {
-        deviceManager.loopManager.getLoopState { [weak self] (_, state) in
-            let endDate = Date()
-            let startDate = endDate.addingTimeInterval(.minutes(-20))
-            let threshold = HKQuantity(unit: GlucoseEffectVelocity.unit, doubleValue: 3)
-
-            let filteredInsulinCounteractionEffects = state.insulinCounteractionEffects.filterDateRange(startDate, endDate)
+        let now = Date()
+        let startDate = now.addingTimeInterval(.minutes(-20))
 
+        deviceManager.glucoseStore.getGlucoseSamples(start: startDate, end: nil) { [weak self] (result) -> Void in
             DispatchQueue.main.async {
-                // at least 3 insulin counteraction effects are required to calculate the average
-                guard filteredInsulinCounteractionEffects.count >= 3,
-                    let averageInsulinCounteractionEffect = filteredInsulinCounteractionEffects.average(unit: GlucoseEffectVelocity.unit) else
-                {
+                switch result {
+                case .failure:
                     self?.shouldDisplayAccurateCarbEntryWarning = false
-                    return
+                case .success(let samples):
+                    let filteredSamples = samples.filterDateRange(startDate, now)
+                    guard let startSample = filteredSamples.first, let endSample = filteredSamples.last else {
+                        self?.shouldDisplayAccurateCarbEntryWarning = false
+                        return
+                    }
+                    let duration = endSample.startDate.timeIntervalSince(startSample.startDate)
```

Comment:
I think you need a `guard duration > 0 else` statement here. This could happen if there is exactly one sample (or potentially two samples with the same `startDate`, which is theoretically possible if you enter a manual glucose sample that has exactly the same time as a CGM sample that backfills slightly later). Without the guard, the `velocity` calculation below will crash.

Reply:
You could actually guard against the `>= .minutes(12)` from below.